### PR TITLE
fix: input

### PR DESCRIPTION
**Pull request by Jannis25**
not tested yet

### CODE REVIEW

File: BPBackendDjango/BPBackendDjango/Views/userviews.py [hunk 1]
```diff
@@ -267,6 +287,15 @@ def post(self, request, *args, **kwargs):
                 }
             return Response(data)
 
+        #check if username is allowed
+        if (not all(c in ALLOWED for c in req_data['username'])) or len(req_data['username']) < MIN_USERNAME_LENGTH or str(req_data['username']).startswith(' '):
```
Comment (Laennart) [inline BPBackendDjango/BPBackendDjango/Views/userviews.py:291]: Here, a check for the maximum username length is missing

File: BPBackendDjango/BPBackendDjango/Views/userviews.py [hunk 2]
```diff
@@ -358,6 +387,15 @@ def post(self, request, *args, **kwargs):
             }
             return Response(data)
 
+        #check if arguments are fine
+        if (FIRST_NAME_LENGTH < len(req_data['first_name'])) or (LAST_NAME_LENGTH < len(req_data['last_name'])) or (EMAIL_LENGTH < len(req_data['email_address'])):
+            data = {
+                'success': False,
+                'description': 'One argument is too long.',
```
Comment (Laennart) [inline BPBackendDjango/BPBackendDjango/Views/userviews.py:394]: This description is very broad. Maybe you could split it up in separate checks.

File: BPBackendDjango/BPBackendDjango/Views/userviews.py [hunk 3]
```diff
@@ -1075,6 +1113,24 @@ def post(self, request, *args, **kwargs):
...
+        #check if length is fine
+        if USERNAME_LENGTH < len(req_data['username']) or (not all(c in ALLOWED for c in req_data['username'])) or len(req_data['username']) < MIN_USERNAME_LENGTH or str(req_data['username']).startswith(' '):
+            data = {
+                    'success': False,
+                    'description': 'username too long',
+                    'data': {}
+                }
+            return Response(data)
+
+        #check if first symbol ist space
```
Comment (Laennart) [inline BPBackendDjango/BPBackendDjango/Views/userviews.py:1125]:
```suggestion
        #check if first symbol is space
```

File: BPBackendDjango/BPBackendDjango/Views/userviews.py [hunk 4]
```diff
@@ -1395,6 +1460,16 @@ def post(self, request, *args, **kwargs):
                     'data': {}
                 }
             return Response(data)
+
+        #check length
+        if TELEPHONE_LENGTH < len(req_data['telephone']):
+            data = {
+                    'success': False,
+                    'description': 'Token is not valid',
```
Comment (Laennart) [inline BPBackendDjango/BPBackendDjango/Views/userviews.py:1468]: I think this description is not the correct one

File: BPBackendDjango/BPBackendDjango/Views/userviews.py [hunk 5]
```diff
@@ -1345,6 +1401,15 @@ def post(self, request, *args, **kwargs):
                     'data': {}
                 }
             return Response(data)
+
+        #check length
+        if (len(req_data['street']) > STREET_LENGTH) or (len(req_data['postal_code']) > POSTAL_CODE_LENGTH) or (len(req_data['country']) > COUNTRY_LENGTH) or (len(req_data['city']) > CITY_LENGTH) or (len(req_data['house_nr']) > H_NR_LENGTH) or (len(req_data['address_add']) > ADDRESS_ADD_LENGTH):
```
Comment (Laennart) [inline BPBackendDjango/BPBackendDjango/Views/userviews.py:1406]: Split this check into separate checks to provide better feedback. Maybe just create a method to check an argument for a given length and another one to create a response with a fitting description.

Comment (Laennart) [reply BPBackendDjango/BPBackendDjango/Views/userviews.py:1406]: Then you can reuse these methods above.

Comment (Laennart) [reply BPBackendDjango/BPBackendDjango/Views/userviews.py:1406]: & on other occasions where lengths are checked

**Review by Laennart — CHANGES_REQUESTED** [review]

File: BPBackendDjango/BPBackendDjango/Views/userviews.py (hunk 1 above)
Comment (Jannis25) [reply BPBackendDjango/BPBackendDjango/Views/userviews.py:291]: no, if username is too long serializer is not valid

Comment (Laennart) [reply BPBackendDjango/BPBackendDjango/Views/userviews.py:291]: what do you mean with serializer?

File: BPBackendDjango/BPBackendDjango/Views/userviews.py [hunk 6]
```diff
@@ -358,6 +391,29 @@ def post(self, request, *args, **kwargs):
...
+            data = {
+                'success': False,
+                'description': 'first name is too long.',
+                'data': {}
+            }
+            return Response(data)
+        if not check_input_length(req_data['email_address'], EMAIL_LENGTH):
+            data = {
+                'success': False,
+                'description': 'first name is too long.',
```
Comment (Laennart) [inline BPBackendDjango/BPBackendDjango/Views/userviews.py:412]: These descriptions are wrong.
As I mentioned in my first review you could create a method like this:
```python
def build_input_length_response(attr_name):
	data = {
		“success“: False,
		“description“: f“{attr_name} too long“,
		“data“: {}	
	}
	return Response(data)
 ```

Then you can use it like this: 
```python
if not check_input_length(req_data['email_address'], EMAIL_LENGTH):
	return build_input_length_response("Email address")
```

That would be way more readable and less duplicate code.

**Review by Laennart — CHANGES_REQUESTED** [review]

**Review by Laennart — APPROVED** [review]
Very nice

**Pull request merged**